### ISSUE TITLE
Unable to Log Gradient Histogram for Weights in MaskedEdgeAttention while Training DialogueGCN

### ISSUE THREAD

**Issue by BorisPolonsky** [SenticNet/conv-emotion#28]
**Describe the bug**
The gradients with respect to weights in `MaskedEdgeAttention` were not properly logged in Tensorboard during training. 
When executing the following line
 https://github.com/SenticNet/conv-emotion/blob/9436e202096b66f025fe36e7059469aeaebe84e7/DialogueGCN/train_IEMOCAP.py#L159
`param[1].grad` will be `None` for the weights in `MaskedEdgeAttention` (which I assume are trainable) instead of a Tensor. These weights include:
- att_model.matchatt.transform.weight
- att_model.matchatt.transform.bias 
- att_model.simpleatt.scalar.weight
- att_model.att.weight
- att_model.att.w_k.weight
- att_model.att.w_k.bias
- att_model.att.w_q.weight found
- att_model.att.w_q.bias found
- att_model.att.proj.weight
- att_model.att.proj.bias

**To Reproduce**
1. Add breakpoint here
 https://github.com/SenticNet/conv-emotion/blob/9436e202096b66f025fe36e7059469aeaebe84e7/DialogueGCN/train_IEMOCAP.py#L159
1. Execute `python train_IEMOCAP.py --base-model 'LSTM' --graph-model --nodal-attention --dropout 0.4 --lr 0.0003 --batch-size 32 --class-weight --l2 0.0 --no-cuda --tensorboard --epochs 1` in debug mode for this test

**Expected behavior**
`param[1].grad` shall not return `None` for at least some of the weights described above.

**Screenshots**

**Desktop (please complete the following information):**
 - OS: Ubuntu 18.04
 - Environment: Python3, Pytorch 1.3

**Smartphone (please complete the following information):**

**Additional context**
I substituted "native" tensorboard support introduced in Pytorch 1.2 for  tensorboardX by replacing this line
https://github.com/SenticNet/conv-emotion/blob/9436e202096b66f025fe36e7059469aeaebe84e7/DialogueGCN/train_IEMOCAP.py#L232
with
```
from torch.utils.tensorboard import SummaryWriter`
```
In this case  
https://github.com/SenticNet/conv-emotion/blob/9436e202096b66f025fe36e7059469aeaebe84e7/DialogueGCN/train_IEMOCAP.py#L159
will raise an Error as it expected a Tensor for gradient instead of `None`.
The equivalent Dockerfile for my container environment is listed as below:
```
FROM pytorch/pytorch:1.3-cuda10.1-cudnn7-devel
RUN conda install -y pandas
RUN pip install torch-geometric torch-sparse torch-scatter torch-cluster
```

**Comment by Zhaobailiang** [comment]
hello, can you help me solve a question? You are so good to run the code.
in my environment:
OS: Ubuntu 18.04
Environment: Python3.6, Pytorch 1.3-cuda10.0-cudnn7,torch-geometric1.3.2
But when I pip install torch-sparse torch-scatter torch-cluster, 
ERROR: Failed building wheel for torch-sparse/torch-scatter/torch-cluster.
I really do not know how to get over that problem.

**Comment by BorisPolonsky** [comment]
> hello, can you help me solve a question? You are so good to run the code.
> in my environment:
> OS: Ubuntu 18.04
> Environment: Python3.6, Pytorch 1.3-cuda10.0-cudnn7,torch-geometric1.3.2
> But when I pip install torch-sparse torch-scatter torch-cluster,
> ERROR: Failed building wheel for torch-sparse/torch-scatter/torch-cluster.
> I really do not know how to get over that problem.

Hello there,
If you are attempting to install in container, the key is to install in the container from `pytorch/pytorch:1.3-cuda10.1-cudnn7-devel` image **WITH GPU SUPPORT** enabled (l.e. by adding the `--runtime nvidia` parameter for nvidia-docker2 or `--gpus all` for the latest version of nvidia-docker plugin, then execute these command manually in the allocated tty:
```
conda install -y pandas
pip install torch-geometric torch-sparse torch-scatter torch-cluster
```
This is because when executing `pip install torch-sparse`, the pip will first judge if you have CUDA support installed, if not, you won't have the CUDA related part of the code installed, hence you will get the error. 

I personally don't like such installation design of torch-sparse. Since I haven't figure out how to expose GPU for `docker build` command, I suggest that you go for the manual approach above, which worked for me.

Keep in mind that you have to run from images with tags ends with `cudnn7-devel` instead of `cudnn7-devel`, otherwise torch-sparse will refuse to build as there's no CUDA related command (e.g. `nvcc`) in containers with tag ends with`-runtime`. 

If this doesn't not solve your problem, I suggest that you file a new issue instead.

**Comment by Zhaobailiang** [comment]
hello, thanks for your help.
 I  installed on the host directly by anaconda. And I try this code: 
torch.cuda.is_available()
![image](https://user-images.githubusercontent.com/38367601/72075001-dfd45180-332d-11ea-9e8d-84e63c7e0ac1.png)
It feedback true. That means torch can find cuda.  But I still can't pip install torch-sparse. Can you know  the reason that I can't pip successfully?

**Comment by Zhaobailiang** [comment]
![image](https://user-images.githubusercontent.com/38367601/72075132-1b6f1b80-332e-11ea-8854-f2b0fc3274ef.png)

**Comment by BorisPolonsky** [comment]
# Case Closed, Not a Bug
## Brief
I took time to read the code just now and concluded that this is not a bug. Hence I shall close this issue.

## Reason & Workaround
The reason is that in `MaskedEdgeAttention`, the constructor defines all possible sub layers regardless of the `attn_mode` (e.g. A redundant `matchatt` field defined as its sub module yet not needed in `forward()` pass in case `attn_mode == 1` is defined), which leads to weights in those redundant modules to be enumerated in the following line
https://github.com/SenticNet/conv-emotion/blob/9436e202096b66f025fe36e7059469aeaebe84e7/DialogueGCN/train_IEMOCAP.py#L158

In case of using official support for tensorboard introduced in new version of pytorch, a simple workaround will be checking if the gradient wrt to the Tensor is None or not before calling `add_histogram` to eliminate corresponding exception.
